bindings/profilers/wall.cc

```diff
@@ -14,224 +14,409 @@
-             TranslateTimeProfileNode(profile->GetTopDownRoot()));
+  static void HandleProfilerSignal(int signal, siginfo_t* info, void* context);
+
+  // Protects the process wide state below.
+  static std::mutex mutex_;
+  static int use_count_;
+  static bool installed_;
+  static struct sigaction old_handler_;
+  using HandlerFunc = void (*)(int, siginfo_t*, void*);
+  static std::atomic<HandlerFunc> old_handler_func_;
+};
+
+std::mutex SignalHandler::mutex_;
+int SignalHandler::use_count_ = 0;
+struct sigaction SignalHandler::old_handler_;
+bool SignalHandler::installed_ = false;
+std::atomic<SignalHandler::HandlerFunc> SignalHandler::old_handler_func_;
+
+void SignalHandler::HandleProfilerSignal(int sig,
+                                         siginfo_t* info,
+                                         void* context) {
+  auto old_handler = old_handler_func_.load(std::memory_order_relaxed);
+
+  if (!old_handler) {
+    return;
+  }
+
+  auto isolate = Isolate::GetCurrent();
+  WallProfiler* prof = g_profilers.GetProfiler(isolate);
+
+  if (!prof) {
+    // no profiler found for current isolate, just pass the signal to old
+    // handler
+    old_handler(sig, info, context);
+    return;
+  }
+
+  // Check if sampling is allowed
+  if (!prof->collectSampleAllowed()) {
+    return;
   }
+
+  auto time_from = Now();
+  old_handler(sig, info, context);
+  auto time_to = Now();
+  prof->PushContext(time_from, time_to);
+}
 #else
-  Nan::Set(js_profile,
-           Nan::New<String>("topDownRoot").ToLocalChecked(),
-           TranslateTimeProfileNode(profile->GetTopDownRoot()));
+class SignalHandler {
+ public:
+  static void IncreaseUseCount() {}
+  static void DecreaseUseCount() {}
+};
 #endif
-  Nan::Set(js_profile,
-           Nan::New<String>("startTime").ToLocalChecked(),
-           Nan::New<Number>(profile->GetStartTime()));
-  Nan::Set(js_profile,
-           Nan::New<String>("endTime").ToLocalChecked(),
-           Nan::New<Number>(profile->GetEndTime()));
-  return js_profile;
-}
+}  // namespace
 
-WallProfiler::WallProfiler(int interval) : samplingInterval(interval) {}
+LabelSetsByNode WallProfiler::GetLabelSetsByNode(CpuProfile* profile,
+                                                 ContextBuffer& contexts) {
+  LabelSetsByNode labelSetsByNode;
 
-WallProfiler::~WallProfiler() {
-  Dispose();
-}
+  auto sampleCount = profile->GetSamplesCount();
+  if (contexts.empty() || sampleCount == 0) {
+    return labelSetsByNode;
+  }
+
+  auto isolate = Isolate::GetCurrent();
+  // auto labelKey = Nan::New<String>("label").ToLocalChecked();
+
+  auto contextIt = contexts.begin();
+
+  // deltaIdx is the offset of the sample to process compared to current
+  // iteration index
+  int deltaIdx = 0;
+
+  // skip first sample because it's the one taken on profiler start, outside of
+  // signal handler
+  for (int i = 1; i < sampleCount; i++) {
+    // Handle out-of-order samples, hypothesis is that at most 2 consecutive
```

Comment:
@nsavoire Hi, I want to know why this happened and why do you need to deal with this situation ? I read the V8 source code and found that the timestamp of samples is incremental (test/cctest/test-cpu-profiler.cc).

```c++
  uint64_t end_time = profile->GetEndTime();
  uint64_t current_time = profile->GetStartTime();
  CHECK_LE(current_time, end_time);
  for (int i = 0; i < profile->GetSamplesCount(); i++) {
    CHECK(profile->GetSample(i));
    uint64_t timestamp = profile->GetSampleTimestamp(i);
    // <=
    CHECK_LE(current_time, timestamp);
    CHECK_LE(timestamp, end_time);
    current_time = timestamp;
  }
```